### PR TITLE
multi: implement waiting and wakeup using winsock events

### PR DESCRIPTION
**Pull request by mback2k**
_Updated PR description:_

This is the 3rd attempt to implement multi-wait and wake-up based upon WinSock events instead of socketpairs.

The 2nd attempt fixed issues with `FD_WRITE` and `FD_CLOSE` not being detected on re-entry, this 3rd attempt fixes issues with `FD_WRITE` not being reset on re-entry.

Ref: #5397
Ref: #5631
Ref: #5634

_Original PR description:_

This is an attempt to restructure the code in Curl_multi_wait
in such a way that less syscalls are made by removing individual
calls to Curl_socket_check via SOCKET_READABLE/SOCKET_WRITABLE.

Bug: #6146

### CODE REVIEW

**Comment by mback2k** [comment]
@rcombs @jay @bagder please take a look at my first attempt to fix/improve #6146 based upon https://github.com/curl/curl/issues/6146#issuecomment-732069116.

**Comment by mback2k** [comment]
Seems like this currently breaks test 1564. I will work on this (hopefully) this evening.

File: lib/multi.c [hunk 1]
```diff
@@ -1385,10 +1352,8 @@ static CURLMcode Curl_multi_wait(struct Curl_multi *multi,
     }
   }
 
-#ifndef USE_WINSOCK
   if(ufds_malloc)
     free(ufds);
-#endif
   if(ret)
     *ret = retcode;
   if(!extrawait || nfds)
```
Comment (mback2k) [inline lib/multi.c:1359]: This condition needs to be updated to also be true if just the WinSock event was triggered.

File: lib/multi.c [hunk 2]
```diff
@@ -1274,30 +1248,32 @@ static CURLMcode Curl_multi_wait(struct Curl_multi *multi,
 #endif
 
   if(nfds) {
```
Comment (mback2k) [inline lib/multi.c:1250]: This condition needs to be updated to also be true if just the WinSock event is being waited on.

**Comment by mback2k** [comment]
@tpodom would you mind helping test the performance with this PR, related to #6146?

**Comment by tpodom** [comment]
@mback2k sorry for the delay getting back to you. The timeout delay still occurs. I see the pollrc = 0 but then the WSAWaitForMultipleEvents delays for 1000 ms.

**Comment by tpodom** [comment]
I did a bit more playing around and I understand why I'm not hitting the WSAEWOULDBLOCK error.  The select checks either in the multi-wait or the socket check prior to send that the socket is ready and if not, the send is not attempted. Since the send is never attempted to trigger WSAEWOULDBLOCK, the FD_WRITE is never signaled for the event so the WSAWaitForMultipleEvents just delays.

I played around with moving the polling after the transfer call and I do see a couple of times where the would block is hit on the send if I modify the send to ignore the socket check but then the timeout_ms delay kicks in and each iteration that gives it enough time to clear up to no longer block.

I think what would really be needed is to set some flag when the WSAEWOUDLBLOCK occurs and then check if that's been set before calling WSAWaitForMultipleEvents. That would require ignoring the socket readiness before send though in transfer.c:
```
  /* If we still have writing to do, we check if we have a writable socket. */
  if ((k->keepon & KEEP_SEND) && (select_res & CURL_CSELECT_OUT)) {
```

Basically, try to replicate the logic as described by MS:

"Therefore, an application can assume that sends are possible starting from the first FD_WRITE network event setting and lasting until a send returns WSAEWOULDBLOCK. After such a failure the application will find out that sends are again possible when an FD_WRITE network event is recorded and the associated event object is set."

**Comment by mback2k** [comment]
> Basically, try to replicate the logic as described by MS:
> 
> "Therefore, an application can assume that sends are possible starting from the first FD_WRITE network event setting and lasting until a send returns WSAEWOULDBLOCK. After such a failure the application will find out that sends are again possible when an FD_WRITE network event is recorded and the associated event object is set."

Makes sense. So an actual fix would require general behavior changes in curl outside of `multi.c`. One question would be: could this be the behavior regardless of platform or would we need to adjust the behavior just on Windows (with WinSock)? cc @bagder @jay

**Comment by tpodom** [comment]
I was thinking about this a bit more and what I'm thinking may not really work at all since I don't have a good grasp on how all of this fits together but what about something along the following:

Curl_send_plain returns CURLE_AGAIN which is handled in Curl_write.  Instead of just returning 0, what if Curl_write updates the connection  to indicate the last write failed with CURLE_AGAIN.

Then, in Curl_multi_wait, only perform the WSAWaitForMultipleEvents if there is no data to write (can we know that in Curl_multi_wait?) or if the write socket hit CURLE_AGAIN. Because if we have data to write and have not hit CURLE_AGAIN then Windows says we should assume we can keep writing.

Then, for the socket check in Curl_readwrite, could Curl_multi_wait set the conn->cselect_bits so the check is skipped, or could Curl_poll be updated to skip the select unless we previously hit CURLE_AGAIN?

**Comment by mback2k** [comment]
@tpodom thanks for the input. I may have found an easier way to reset `FD_WRITE` by just sending zero bytes before trying to wait on a socket for write readiness again. Would you mind testing the new implementation in this updated PR? cc @jay @rcombs

**Comment by jay** [comment]
Using the same test from #6146:
~~~
master (796c068 2020-12-14)
Transfer rate: 3992 KB/sec (52429310 bytes in 13 seconds)
Transfer rate: 4124 KB/sec (52429310 bytes in 12 seconds)
Transfer rate: 4148 KB/sec (52429310 bytes in 12 seconds)

mback2k:multi-consolidate-syscalls (c502f6a 2020-12-15)
Transfer rate: 4107 KB/sec (52429310 bytes in 12 seconds)
Transfer rate: 4059 KB/sec (52429310 bytes in 13 seconds)
Transfer rate: 4141 KB/sec (52429310 bytes in 12 seconds)
~~~~

File: lib/multi.c [hunk 3]
```diff
@@ -1159,22 +1173,38 @@ static CURLMcode Curl_multi_wait(struct Curl_multi *multi,
...
+#ifdef USE_WINSOCK
+          mask |= FD_WRITE|FD_CONNECT|FD_CLOSE;
+          send(s, NULL, 0, 0); /* reset FD_WRITE */
+#endif
+          ufds[nfds].fd = s;
           ufds[nfds].events = POLLOUT;
           ++nfds;
-          s = sockbunch[i];
         }
         if(s == CURL_SOCKET_BAD) {
```
Comment (mback2k) [inline lib/multi.c:1198]: @bagder while working on this it came to my mind if this check shouldn't be at the top of the loop to avoid counting one invalid socket?

**Comment by tpodom** [comment]
@mback2k sorry for the delay getting back to you.  The latest change looks to be working great for me.  I tested with auto tuning both enabled and disabled on the server and did not experience any significant delays and the time to upload was consistent.  Good find on using the 0 byte send to trigger the FD_WRITE!

**Comment by mback2k** [comment]
@tpodom no problem, thanks for testing. @jay thanks for verifying the performance, would you mind reviewing the PR?

Can I also get you @Togtja, @ngg, @rmja, @tmkk, @RadMission to test this PR regarding up- and download performance issues you experienced with past iterations?

**Comment by tmkk** [comment]
Run the same test from #5631

Without this PR:
```
% ./a.exe
192.168.1.160:80...
* Connected to 192.168.1.160 (192.168.1.160) port 80 (#0)
> POST / HTTP/1.1
Host: 192.168.1.160
Accept: */*
Content-Length: 1048576
Content-Type: application/x-www-form-urlencoded

Filled 65536/1048576 bytes, 5.002 msec elapsed
Filled 131072/1048576 bytes, 5.002 msec elapsed
Filled 196608/1048576 bytes, 8.000 msec elapsed
Filled 262144/1048576 bytes, 8.999 msec elapsed
Filled 327680/1048576 bytes, 8.999 msec elapsed
Filled 393216/1048576 bytes, 8.999 msec elapsed
Filled 458752/1048576 bytes, 9.998 msec elapsed
Filled 524288/1048576 bytes, 9.998 msec elapsed
Filled 589824/1048576 bytes, 9.998 msec elapsed
Filled 655360/1048576 bytes, 10.998 msec elapsed
Filled 720896/1048576 bytes, 10.998 msec elapsed
Filled 786432/1048576 bytes, 11.998 msec elapsed
Filled 851968/1048576 bytes, 11.998 msec elapsed
Filled 917504/1048576 bytes, 11.998 msec elapsed
Filled 983040/1048576 bytes, 13.008 msec elapsed
Filled 1048576/1048576 bytes, 13.008 msec elapsed
* We are completely uploaded and fine
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
```
With this PR:
```
% ./a.exe
192.168.1.160:80...
* Connected to 192.168.1.160 (192.168.1.160) port 80 (#0)
> POST / HTTP/1.1
Host: 192.168.1.160
Accept: */*
Content-Length: 1048576
Content-Type: application/x-www-form-urlencoded

Filled 65536/1048576 bytes, 4.996 msec elapsed
Filled 131072/1048576 bytes, 4.996 msec elapsed
Filled 196608/1048576 bytes, 7.995 msec elapsed
Filled 262144/1048576 bytes, 8.995 msec elapsed
Filled 327680/1048576 bytes, 8.995 msec elapsed
Filled 393216/1048576 bytes, 8.995 msec elapsed
Filled 458752/1048576 bytes, 9.994 msec elapsed
Filled 524288/1048576 bytes, 9.994 msec elapsed
Filled 589824/1048576 bytes, 9.994 msec elapsed
Filled 655360/1048576 bytes, 10.994 msec elapsed
Filled 720896/1048576 bytes, 10.994 msec elapsed
Filled 786432/1048576 bytes, 10.994 msec elapsed
Filled 851968/1048576 bytes, 11.993 msec elapsed
Filled 917504/1048576 bytes, 11.993 msec elapsed
Filled 983040/1048576 bytes, 12.992 msec elapsed
Filled 1048576/1048576 bytes, 12.992 msec elapsed
* We are completely uploaded and fine
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
```
No regression observed, looks good to me.

File: lib/multi.c [hunk 4]
```diff
@@ -374,6 +374,11 @@ struct Curl_multi *Curl_multi_handle(int hashsize, /* socket hash */
   multi->max_concurrent_streams = 100;
   multi->ipv6_works = Curl_ipv6works(NULL);
 
+#ifdef USE_WINSOCK
+  multi->wsa_event = WSACreateEvent();
+  if(multi->wsa_event == WSA_INVALID_EVENT)
+    goto error;
+#else
```
Comment (jay) [inline lib/multi.c:381]: Why is this independent of ENABLE_WAKEUP?

File: lib/multi.c [hunk 5]
```diff
@@ -1223,6 +1306,33 @@ static CURLMcode Curl_multi_wait(struct Curl_multi *multi,
...
+            if(bitmap & (GETSOCK_READSOCK(i) | GETSOCK_WRITESOCK(i))) {
+              wsa_events.lNetworkEvents = 0;
+              if(WSAEnumNetworkEvents(sockbunch[i], NULL, &wsa_events) == 0) {
+                if(ret && pollrc <= 0 && wsa_events.lNetworkEvents != 0)
+                  retcode++;
+              }
+              WSAEventSelect(sockbunch[i], multi->wsa_event, 0);
+            }
+            else
+              break;
```
Comment (jay) [inline lib/multi.c:1327]: Why are you breaking out of this, isn't multi poll supposed to return the total number of sockets on which events have happened, and wouldn't that require checking all the sockets?

File: lib/multi.c (hunk 4 above)
Comment (mback2k) [reply lib/multi.c:381]: Because the event is used for normal socket waiting as well. This single event is used for all waiting purposes.

File: lib/multi.c (hunk 5 above)
Comment (mback2k) [reply lib/multi.c:1327]: Yes, but the bitmap tells how many sockets we were supposed to check in the first place. So this is just the same condition as for the loops in the setup of the waiting at the beginning of the function.

Comment (mback2k) [reply lib/multi.c:1327]: Wait a moment, you are right. This is not exactly the same condition as in the setup of the function. Also see my comment to @bagder above that there is another condition in the setup.

**Comment by ngg** [comment]
Test 1564 (a wakeup test) has a part which is disabled on Windows: https://github.com/curl/curl/blob/6cd066f64b9165c1b1f3243844ea32b1f1ba42f5/tests/libtest/lib1564.c#L112-L134
I originally disabled it because the socketpair on Windows was working strange, with these changes I think it should work.
Can you check if it can be enabled?

File: lib/multi.c (hunk 3 above)
Comment (jay) [reply lib/multi.c:1198]: > shouldn't be at the top of the loop to avoid counting one invalid socket?

it seems s is only set to a valid socket if the bitmap for readable/writable was set for that socket, and so at this point we're basically checking whether that has happened, but I don't know why it's breaking if the socket is invalid I would expect it to continue checking the bitmap.

**Comment by mback2k** [comment]
Thanks, I hope to have the time to work on the comments given above soon.

File: lib/multi.c (hunk 4 above)
Comment (mback2k) [reply lib/multi.c:381]: @jay can I consider this comment solved then?

Comment (jay) [reply lib/multi.c:381]: yes

**Comment by mback2k** [comment]
Rebased and enable the test part as suggested by @ngg, will be interesting to see what CI will show.

File: lib/multi.c (hunk 3 above)
Comment (mback2k) [reply lib/multi.c:1182]: @bagder would you mind explaining why the bitmap is not further checked if readable/writable was set for a socket?

**Comment by mback2k** [comment]
TODO during PR merge: add comments above the bitmap conditions to explain them better.

**Comment by jay** [comment]
I'm fine with this with the exception of the comment I made about the loop in multi_wait. @bagder will have to help you with that because I don't understand what's going on there. Two suggestions, squash commits together (it looks like the middle 2 could be squashed into the consolidate commit?) and though this isn't technically a feature wait until the next feature window because I think it should sit in development builds for the longer period.

**Comment by mback2k** [comment]
Sure, I plan on waiting for @bagder on this and the next feature window. Regarding squashing: I will do this during merge, at the moment I would like to keep things separate.

**Comment by mback2k** [comment]
@bagder: @jay is referring to the loop conditions discussed in https://github.com/curl/curl/pull/6245#discussion_r546570303 and in https://github.com/curl/curl/pull/6245#discussion_r546543745. Would you be able to help us here by explaining why the original loop is stopped if no bit is set for a socket and no further sockets are checked?

@jay: just FYI, the original logic implemented in the first loop discussed in https://github.com/curl/curl/pull/6245#discussion_r546570303 is not changed by this PR and just replicated in https://github.com/curl/curl/pull/6245#discussion_r546543745. Therefore I don't think the discussion is critical to a change in this PR, but of course an explanation would be helpful anyway.

**Comment by mback2k** [comment]
@bagder sorry, for me the links to the review comments above still work, but anyway, we are talking about the break condition in this loop:
https://github.com/curl/curl/blob/b899cb08c66055cb798d8109de0e19e371fe440a/lib/multi.c#L1026-L1042

Which is then replicated here in a slightly different way a second time as part of this PR, but still follows the same logic of breaking if neither `GETSOCK_READSOCK` or `GETSOCK_WRITESOCK` is set:
https://github.com/mback2k/curl/blob/4f0b65a5f990221acfebcfe69b12378f1ab041db/lib/multi.c#L1303-L1316

We are wondering why no further checking of additional sockets is done in that case and therefore why the condition `while(data) {` is not enough for the loop to terminate itself.

**Comment by bagder** [comment]
Those loops iterate from 0 to `MAX_SOCKSPEREASYHANDLE` and since we don't know how many sockets were filled in the array, it instead breaks when it comes to the first entry that isn't filled in with a legitimate socket.

**Comment by mback2k** [comment]
This is what I suspected. Any other questions @jay?

**Comment by jay** [comment]
I reviewed this a while ago and I don't remember specifics. Also the 'view changes' by that comment I made cannot be found. I was commenting on this, which you linked to
~~~c
          for(i = 0; i < MAX_SOCKSPEREASYHANDLE; i++) {
            if(bitmap & (GETSOCK_READSOCK(i) | GETSOCK_WRITESOCK(i))) {
              wsa_events.lNetworkEvents = 0;
              if(WSAEnumNetworkEvents(sockbunch[i], NULL, &wsa_events) == 0) {
                if(ret && pollrc <= 0 && wsa_events.lNetworkEvents != 0)
                  retcode++;
              }
              WSAEventSelect(sockbunch[i], multi->wsa_event, 0);
            }
            else {
              /* break on entry not checked for being readable or writable */
              break;
~~~
How do we know that if firstsocket is not read/write then secondsocket isn't either? Also max socks is 5 so I'm not sure why that is, but basically my point was why wouldn't you keep on and check all of them just because the first one is not read/write. Doesn't mean it's bad. Unless I misunderstand.

**Comment by mback2k** [comment]
Yes, this is how I understood your comments / questions, but to reiterate: `if(bitmap & (GETSOCK_READSOCK(i) | GETSOCK_WRITESOCK(i))) {` is checking whether we are supposed to test the socket for readability or writability in the first place, not if it actually is ready. The same condition - just split into two - is used all over `multi.c`. If I understood @bagder correctly, a valid socket entry in `sockbunch` needs to have one of these flags set. So it is like a null-terminated string, once we reach a socket entry without one of the flags set, we stop iterating, right?

And just for reference, these are the review comments I am referring to which are still perfectly visible:
image
image

**Comment by jay** [comment]
So, as @bagder has described it, when neither flag is set in the bitmap on the corresponding socket in the array then no sockets further in the array will have either flag. If that is the case I don't think further discussion is necessary.

**Comment by mback2k** [comment]
Thanks @jay! So I will rebase this PR soon and then would like to ask you Windows guys @jay @MarcelRaad for some final feedback/testing and review/approval for this PR to be merged during the next feature window. I also reached out on Twitter to ask other curl Windows developers to test the performance impact of this PR, but unfortunately no one showed up here so far.

File: lib/multi.c (hunk 5 above)
Comment (mback2k) [reply lib/multi.c:1314]: Resolved, see https://github.com/curl/curl/pull/6245#issuecomment-797323267.

File: lib/multi.c (hunk 3 above)
Comment (mback2k) [reply lib/multi.c:1183]: Resolved, see https://github.com/curl/curl/pull/6245#issuecomment-797323267.

**Comment by MarcelRaad** [comment]
Looks good to me in general. Will try to do some testing on the weekend.

**Comment by jay** [comment]
Using the same [postupload](https://gist.github.com/jay/daac01682f17c2dc213e) example to the closest test server (ny2 for me) the results are very similar.

~~~
master (f83d4ea 2021-03-14)
Transfer rate: 4123 KB/sec (52429310 bytes in 12 seconds)
Transfer rate: 3984 KB/sec (52429310 bytes in 13 seconds)
Transfer rate: 4043 KB/sec (52429310 bytes in 13 seconds)
Transfer rate: 4101 KB/sec (52429310 bytes in 12 seconds)
Transfer rate: 4206 KB/sec (52429310 bytes in 12 seconds)
Transfer rate: 3951 KB/sec (52429310 bytes in 13 seconds)

mback2k:multi-consolidate-syscalls (ca6ce77 2021-03-12)
Transfer rate: 4013 KB/sec (52429310 bytes in 13 seconds)
Transfer rate: 4045 KB/sec (52429310 bytes in 13 seconds)
Transfer rate: 4047 KB/sec (52429310 bytes in 13 seconds)
Transfer rate: 4030 KB/sec (52429310 bytes in 13 seconds)
Transfer rate: 4079 KB/sec (52429310 bytes in 13 seconds)
Transfer rate: 4018 KB/sec (52429310 bytes in 13 seconds)
~~~

master (what multi-consolidate-syscall is currently branched off of) is overall a few KB faster but that is not of significance due to network conditions.

**Comment by mback2k** [comment]
@jay thanks a lot for testing this again! I would really appreciate your approval on this PR then if you feel like it is ready to go in during the next feature window. I will also perform some additional upload *and* download performance tests before merge.

**Review by jay — APPROVED** [review]
ok for next-feature-window. suggest squash whatever you can to make it easier in case we have to revert.

**Review by MarcelRaad — APPROVED** [review]

**Comment by mback2k** [comment]
Rebased to current master in order to get ready for merge during this feature window.

**Comment by jay** [comment]
> Rebased to current master in order to get ready for merge during this feature window.

bugfixes only until the [patch release on the 14th](https://curl.se/mail/lib-2021-04/0030.html)

**Comment by mback2k** [comment]
Rebased again to current master in order to get ready for merge during this feature window.

**Comment by mback2k** [comment]
Rebased again due to previously failing test 1660 which was fixed with 6b97f1311a1454c83ca04665a9eec2673643fd1d.

**Comment by mback2k** [comment]
Thanks everyone! @rcombs @tpodom @tmkk @jay @MarcelRaad
